cmd/etcd-launcher/main.go

```diff
@@ -26,66 +33,137 @@ type envConfig struct {
 	dataDir               string
 	token                 string
 	enableCorruptionCheck bool
+	initialState          string
+}
+
+type etcdCluster struct {
+	config      *envConfig
+	client      *clientv3.Client
+	localClient *clientv3.Client
 }
 
 func main() {
 	// normal workflow, we don't do migrations anymore
-	config, err := getConfigFromEnv()
+	e := &etcdCluster{}
+	err := e.getConfigFromEnv()
 	if err != nil {
 		log.Fatalf("failed to get launcher configuration: %v", err)
 	}
 
+	initialMembers := initialMemberList(e.config.clusterSize, e.config.namespace)
 	// not required, will leave it for now.
-	os.Setenv(initialStateEnvName, "new")
-	os.Setenv(initialClusterEnvName, initialMemberList(config.clusterSize, config.namespace))
+	os.Setenv(initialStateEnvName, e.config.initialState)
+	os.Setenv(initialClusterEnvName, strings.Join(initialMembers, ","))
 
 	log.Print("initializing etcd..")
 	log.Printf("initial-state: %s", os.Getenv(initialStateEnvName))
 	log.Printf("initial-cluster: %s", os.Getenv(initialClusterEnvName))
 
-	err = syscall.Exec(
-		"/usr/local/bin/etcd",
-		etcdCmd(config),
-		os.Environ())
+	// setup and start etcd command
+	cmd := exec.Command("/usr/local/bin/etcd", etcdCmd(e.config)...)
+	cmd.Env = os.Environ()
+	cmd.Stderr = os.Stderr
+	cmd.Stdout = os.Stdout
+	if err = cmd.Start(); err != nil {
+		log.Fatalf("failed to start etcd: %v", err)
+	}
+
+	// if existing, we need to reconcile
+	var healthy bool
+	for i := 0; i < 5; i++ {
```

Comment:
So here you are trying to make sure that the storage is healthy, so you are trying to get valid results in 5 times. In general that's valid of course but what about adding some semantics something similar to this
```
	var (
		tries    int
		maxTries int
                healthy bool
	)

	for tries < maxTries {
		if healthy, err = e.isHealthy(); !healthy || err != nil {
			log.Error("...")
			time.Sleep(500 * time.Milisecond)
			tries++
			continue
		}
		log.Info("...")
		break
	}

      if !healthy || err != nil {
       log.Fatal("failed for ever")
     } 
```

Reply:
Or you can try out this lovely package:
https://godoc.org/k8s.io/apimachinery/pkg/util/wait#Poll